### PR TITLE
all: upgrade minimum supported Go release to 1.22

### PR DESCRIPTION
**Pull request by nsrip-dd**
### What does this PR do?

With Go 1.23 released on 2024-08-13, we drop support for Go 1.21. This
means Go 1.22 is now our minimum supported Go release. Update our CI
files to reflect our current supported Go versions, and remove pre-Go
1.22 workarounds.

In addition to the CI bumps, this PR changes/fixes a few things related to the upgrade:

- Drops the old `rand` from the tracer, now that we can depend on `math/rand/v2`.
- Uses `slices` in the appsec package.
- The `TestWrapHandlerWithResourceNameNoRace` got a new data race. Specifically,
  by bumping our `go.mod` to 1.22, we get the new 1.22 ServeMux behavior internally.
  The 1.22 ServeMux modifies the Request, introducing a race with all the goroutines
  accessing/modifying the request concurrently. Fix it by giving each goroutine its own
  request. I verified that the test still catches the intended bug by reverting the fix from
  #2204 and verifying that the test fails.

### Motivation

Follow our declared Go support policy.

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: .github/workflows/unit-integration-tests.yml [hunk 1]
```diff
@@ -28,7 +28,10 @@ jobs:
         uses: actions/checkout@v3
         with:
           ref: ${{ inputs.ref || github.ref }}
-
+      - name: Setup go
+        uses: actions/setup-go@v5
```
[the remark this answers is not in the document]
Comment (nsrip-dd) [reply .github/workflows/unit-integration-tests.yml:32]: Ack.

**Review by darccio — APPROVED** [review]
LGTM

**Review by eliottness — CHANGES_REQUESTED** [review]
I believe there are small changes to do [here](https://github.com/DataDog/dd-trace-go/blob/c29e8fa1aea7cf0be7c0a885fc5d1424089396c0/contrib/net/http/http_test.go#L320)

**Comment by nsrip-dd** [comment]
> I believe there are small changes to do [here](https://github.com/DataDog/dd-trace-go/blob/c29e8fa1aea7cf0be7c0a885fc5d1424089396c0/contrib/net/http/http_test.go#L320)

Done, thanks for the review @eliottness

**Review by eliottness — APPROVED** [review]

**Pull request merged**